crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -23,17 +84,255 @@ where
+
+        // Update failed_notes with notes that weren't included in successful combination
+        let successful_note_ids = successful.iter().map(|note| note.id()).collect::<BTreeSet<_>>();
+        let newly_failed: Vec<_> = candidate_notes
+            .into_iter()
+            .filter(|input_note| !successful_note_ids.contains(&input_note.note().id()))
+            .map(|input_note| {
+                FailedNote::new(
+                    input_note.into_note(),
+                    TransactionExecutorError::DiscardedDuringRetry,
+                )
+            })
+            .collect();
+        failed_notes.extend(newly_failed);
+
+        Ok(NoteConsumptionInfo::new(successful, failed_notes))
+    }
+
+    /// Validates input notes, transaction inputs, and account inputs before executing the
+    /// transaction with specified notes. Keeps track and returns both successfully consumed notes
+    /// as well as notes that failed to be consumed.
+    ///
+    /// The `source_manager` is used to map potential errors back to their source code. To get the
+    /// most value out of it, use the source manager from the
+    /// [`Assembler`](miden_objects::assembly::Assembler) that assembled the Miden Assembly code
+    /// that should be debugged, e.g. account components, note scripts or transaction scripts. If
+    /// no error-to-source mapping is desired, a default source manager can be passed, e.g.
+    /// [`DefaultSourceManager::default`](miden_objects::assembly::DefaultSourceManager::default).
+    ///
+    /// # Returns:
+    /// - An index into the input notes for the note that failed execution along with the associated
+    ///   error.
+    ///
+    /// # Errors:
+    /// Returns an error if:
+    /// - If required data can not be fetched from the [`DataStore`].
+    /// - If the transaction host can not be created from the provided values.
+    /// - If the execution of the provided program fails on the stage other than note execution.
+    async fn try_execute_notes(
+        &self,
+        account_id: AccountId,
+        block_ref: BlockNumber,
+        notes: InputNotes<InputNote>,
+        tx_args: &TransactionArgs,
+    ) -> Result<TransactionExecutionAttempt, TransactionExecutorError> {
+        if notes.is_empty() {
+            return Ok(TransactionExecutionAttempt::Successful);
+        }
+
+        // TODO: ideally, we should prepare the inputs only once for the while note consumption
+        // check (rather than doing this every time when we try to execute some subset of notes),
+        // but we currently cannot do this because transaction preparation includes input notes;
+        // we should refactor the preparation process to separate input note preparation from the
+        // rest, and then we can prepare the rest of the inputs once for the whole check
+        let (mut host, _, stack_inputs, advice_inputs) =
+            self.0.prepare_transaction(account_id, block_ref, notes, tx_args, None).await?;
```

Comment:
This is not something we should do in this PR, but let's create an issue for this. Basically, we want to skip the step of loading account, chain MMR etc. separately for each note test - but for this we'd need to refactor things slightly.